### PR TITLE
ROS 2 IDL support

### PR DESCRIPTION
**Pull request by snosenzo**
- add ros2idl option to parser
- add ros2idl grammar for idl parsing
- add tests for ros2idl grammar

Open questions:
 - ~~do we want to keep the message paths separated by `::` ? or should I replace those with a `/`?~~
   - type names should match their package resource names, there should be no `::`'s in output ASTs. 
 - ~~Does the IDL parsing also need to perform fixup types? As an extension,  I'm not sure how we want to handle `#includes` or if we just assume that everything needed should be in the message definition string. I imagine everything needed should be in the message definition string.~~
   - won't need to use `#include`s in any kind of way since the messages should be in the file later. 

Checklist:
- [x] Clean up grammar and add comments
- [x] Normalize definitions and update tests
- [x] Support scientific notation without signs
- [x] Support multiple declarations in a single line?

### CODE REVIEW

File: src/ros2idl.ne [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+# unsupported annotations are ignored
+rangeAnnotation -> _ at "range" _ "(" _ "min" assignment _ "," _ "max" assignment _ ")" {% noop %}
+
+commentAnnotation -> _ at "verbatim" _ "(" _ "language" assignment _ "," _ "text" _ %EQ _ STR ")" {% noop %}
+
+keyAnnotation -> _ at "key" {% noop %}
+
+transferModeAnnotation -> _ at "transfer_mode" _ "(" _ %NAME _ ")" {% noop %}
+
+at -> "@" {% noop %}
```
Comment (jtbandes) [inline src/ros2idl.ne:226]: not super familiar with how the lexer stuff works, but why do we have both `at` and `AT`?

File: src/ros2idl.ne [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+} %}
+
+# unsupported annotations are ignored
+rangeAnnotation -> _ at "range" _ "(" _ "min" assignment _ "," _ "max" assignment _ ")" {% noop %}
+
+commentAnnotation -> _ at "verbatim" _ "(" _ "language" assignment _ "," _ "text" _ %EQ _ STR ")" {% noop %}
+
+keyAnnotation -> _ at "key" {% noop %}
+
+transferModeAnnotation -> _ at "transfer_mode" _ "(" _ %NAME _ ")" {% noop %}
```
Comment (jtbandes) [inline src/ros2idl.ne:224]: surprised to see `_` at the beginning of all of these – maybe it should be added at a higher level? or is it here to make these rules more convenient to use elsewhere?

File: src/ros2idl.ne [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+
+const moo = require("moo");
+// Terminal tokens are in all caps
+const lexer = moo.compile({
+  SPACE: {match: /\s+/, lineBreaks: true},
+  DECIMALEXP: /(?:(?:\d+\.\d*)|(?:\d*\.\d+)|(?:[0-9]+))[eE](?:[+|-])?[0-9]+/,
+  DECIMAL: /(?:(?:\d+\.\d*)|(?:\d*\.\d+))/,
+  INTEGER: /[0-9]+/,
+  COMMENT: /(?:\/\/[^\n]*)|(?:\/\*(?:.|\n)+?\*\/)/,
+  HEX_LITERAL: /0x(?:[0-9][a-f][A-F])+?/,
```
Comment (jtbandes) [inline src/ros2idl.ne:50]:
```suggestion
  HEX_LITERAL: /0x[0-9a-fA-F]+/,
```

File: src/ros2idl.ne [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+	return obj;
+}, {});
+
+const moo = require("moo");
+// Terminal tokens are in all caps
+const lexer = moo.compile({
+  SPACE: {match: /\s+/, lineBreaks: true},
+  DECIMALEXP: /(?:(?:\d+\.\d*)|(?:\d*\.\d+)|(?:[0-9]+))[eE](?:[+|-])?[0-9]+/,
+  DECIMAL: /(?:(?:\d+\.\d*)|(?:\d*\.\d+))/,
+  INTEGER: /[0-9]+/,
```
Comment (jtbandes) [inline src/ros2idl.ne:48]: nit
```suggestion
  INTEGER: /\d+/,
```

File: src/ros2idl.ne [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+	return d[1][0];
+}
+%}
+
+# support <import> or "import" includes - just ignored
+importDcl -> _ "#" "include" __ (%STRING | "<" _ %NAME ("/" %NAME):* "." "idl" _ ">") {% noop %}
+
+# constant modules need to be separate from complex modules since they shouldn't mix
+module  -> ((comment|annotation):* _ "module" __ fieldName __ "{" __ (constantDcl):+ __ "}" semi) {% processConstantModule %}
+  | ((comment|annotation):* _ "module" __ fieldName __ "{" __ (structWithAnnotation|module):+ __ "}" semi) {% processComplexModule %}
```
Comment (jtbandes) [inline src/ros2idl.ne:149]: is there a reason that `comment` has to be here at the beginning? can it be up at the `main` level?

Alternatively, I suppose maybe we technically need to support comments occurring anywhere on a line (like `module /*comment*/ { ... }` ... is that something the lexer can remove so we don't have to deal with them in the grammar?

File: src/ros2idl.ne [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+
+
+structWithAnnotation -> (comment|annotation):* struct {% 
+ // ignore annotations on structs because we can only read default value (which doesn't apply)
+ d => {return d[1];}
```
Comment (jtbandes) [inline src/ros2idl.ne:154]: what does "doesn't apply" mean here?
nit
```suggestion
 d => d[1]
```

File: src/ros2idl.ne [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+%}
+
+struct -> _ "struct" __ fieldName __ "{" __ (declaration):+ __ "}" semi {% d => {
```
Comment (jtbandes) [inline src/ros2idl.ne:157]: just guessing that these WS are not actually required?
```suggestion
struct -> _ "struct" __ fieldName _ "{" _ (declaration):+ _ "}" _ semi {% d => {
```

Comment (jtbandes) [inline src/ros2idl.ne:157]: is an empty struct (no declarations) not allowed?

File: src/ros2idl.ne [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+constType -> (
+     _ constKeyword __ numericType __ fieldName floatAssignment _ simple
+   | _ constKeyword __ numericType __ fieldName intAssignment _ simple
+   | _ constKeyword __ stringType __ fieldName stringAssignment _ simple
+   | _ constKeyword __ booleanType __ fieldName booleanAssignment _ simple
+) {% d => {
+	const def = extend(d[0]);
+	const name = def.name;
+	const value = def.value;
+	constantToValueMap[name] = value;
```
Comment (jtbandes) [inline src/ros2idl.ne:239]: is it normal to have side effects in these postprocessing functions? does that ever cause problems if the nearley engine needs to do backtracking? I think it might be more normal to just parse the AST leaving any variable/constant references intact and not worry about filling in the values until a later step.

File: src/ros2idl.ne [hunk 9]
```diff
@@ -0,0 +1,395 @@
...
+INT -> %INTEGER {% join %}
+
+semi -> ";" {% noop %}
+
+
+# Optional whitespace
+_ -> (null | %SPACE) {% noop %}
+
+# Required whitespace
+__ -> %SPACE {% noop %}
```
Comment (jtbandes) [inline src/ros2idl.ne:391]: on [this page](https://nearley.js.org/docs/tokenizers) they mention their tokenizer removes whitespace, but in the moo example below they use a `%ws` token... [this page](https://nearley.js.org/docs/how-to-grammar-good) also mentions the `_` & `__` convention for whitespace... Is it an option for us to handle whitespace purely in the lexing phase and not have to deal with it during parsing? Or do we need it in the grammar for some reason?

File: src/ros2idl.ne [hunk 10]
```diff
@@ -0,0 +1,395 @@
...
+
+SIGNED_INT -> ("+"|"-") INT  {% join %}
+
+INT -> %INTEGER {% join %}
+
+semi -> ";" {% noop %}
+
+
+# Optional whitespace
+_ -> (null | %SPACE) {% noop %}
```
Comment (jtbandes) [inline src/ros2idl.ne:388]: does this work?
```suggestion
_ -> %SPACE:? {% noop %}
```

File: src/ros2idl.ne [hunk 11]
```diff
@@ -0,0 +1,395 @@
...
+SIGNED_FLOAT -> ("+"|"-") FLOAT {% join %}
+
+FLOAT -> (%DECIMAL|%DECIMALEXP) {% join %}
+ | (%DECIMAL "d") {% d => d[0][0].value %}
+ | (INT "d") {% d => d[0][0] %}
+
+
+SIGNED_INT -> ("+"|"-") INT  {% join %}
+
+INT -> %INTEGER {% join %}
```
Comment (jtbandes) [inline src/ros2idl.ne:382]: what's the point of this rule?

File: src/ros2idl.ne [hunk 12]
```diff
@@ -0,0 +1,395 @@
...
+# ALL CAPS return strings rather than objects (terminals)
+
+BOOLEAN -> ("TRUE" | "FALSE") {% join %}
+
+# need to support mutliple adjacent strings as a single string
+STR -> (%STRING _):+  {% d => {
+	return join(d[0].flat(1).filter(d => d !== null));
+}%}
+
+# Not actually used due to needing to parse either an int or float from a string
```
Comment (jtbandes) [inline src/ros2idl.ne:370]: I don't really understand this comment, if it's not used, can we get rid of it?

File: src/ros2idl.ne [hunk 13]
```diff
@@ -0,0 +1,395 @@
...
+			
+		case "octet":
+			return {type: "byte"};
+		
+		case "wchar": 
+			return {type: "char"};
+
+		default:
+			return { type: maybeType } ;
+	}
```
Comment (jtbandes) [inline src/ros2idl.ne:357]: could probably be shorter/cleaner with a map defined at the top of the file

File: src/ros2idl.ne [hunk 14]
```diff
@@ -0,0 +1,395 @@
...
+	let strLength = undefined;
+	if(d[1] !== null) {
+		strLength = getIntOrConstantValue(d[1][3] ? d[1][3][0] : undefined);
+	}
+	return {type: "string", upperBound: strLength};
+} %}
+
+booleanType -> "boolean" {% d => ({type: "bool"}) %}
+
+# order matters here
```
Comment (jtbandes) [inline src/ros2idl.ne:297]: why?

File: src/parse.ros2idl.test.ts [hunk 15]
```diff
@@ -0,0 +1,886 @@
...
+        struct MyAction_Goal {
+          int32 input_value;
+        };
+      };
+    `;
+    expect(() => parse(msgDef, { ros2idl: true })).toThrow(
+      `Could not parse message definition (unexpected end of input): '${msgDef}'`,
+    );
+  });
+  it("cannot parse empty module", () => {
```
Comment (jtbandes) [inline src/parse.ros2idl.test.ts:873]: is this actually a syntax error in IDL or just something we don't support today?

File: src/ros2idl.ne [hunk 16]
```diff
@@ -0,0 +1,395 @@
...
+  | commentAnnotation
+  | keyAnnotation
+  | transferModeAnnotation
+  ) {% d => id(id(d)) %}
+
+defaultAnnotation -> _ at "default" _ "(" _ "value" assignment _ ")" {% d => { 
+	return {defaultValue: d[7].value };
+} %}
+
+# unsupported annotations are ignored
```
Comment (jtbandes) [inline src/ros2idl.ne:217]: I don't see anything here about unsupported annotations (`@` + some word we don't recognize)...how do we ignore them? or is the goal to actually throw errors for cases we aren't sure about since they might change the meaning of the idl in an unknown way? I think the comment could use more detail at least.

File: src/parse.ts [hunk 17]
```diff
@@ -12,12 +12,18 @@ import { Grammar, Parser } from "nearley";
...
+const ROS2IDL_GRAMMAR = Grammar.fromCompiled(ros2idlRules);
 
 export type ParseOptions = {
-  /** Parse message definitions as ROS2. Otherwise, parse as ROS1 */
+  /** Parse message definitions as ROS 2. Otherwise, parse as ROS1 */
   ros2?: boolean;
+  /** Parse message definitions in the ROS 2 subset of IDL.
+   * Ignores ros2 and skipTypeFixup option value.
+   */
+  ros2idl?: boolean;
```
Comment (jtbandes) [inline src/parse.ts:26]: maybe time to change this to `syntax?: "ros1" | "ros2msg" | "ros2idl"` (open to bikeshedding on names)?

Comment (jtbandes) [reply src/parse.ts:26]: On the other hand, I am left wondering why these are all in one `parse()` function. Maybe we should just have `parseRos1Msg`, `parseRos2Msg`, and `parseRos2Idl` functions.

**Review by jtbandes — COMMENTED** [review]
Looks cool. Read through some parts of the grammar and posted some questions (many due to my ignorance of nearley). Have you tried running this on all the autoware idls yet?

File: src/parse.ts (hunk 17 above)
Comment (defunctzombie) [reply src/parse.ts:26]: I too wonder this. In my mind all the downstream places where we parse a scheme we have an encoding that would tell us which type of parser to use.

**Comment by defunctzombie** [comment]
> do we want to keep the message paths separated by :: ? or should I replace those with a /?

What is the correct format for names from idl? For example in protobuf we keep the period separator because that is idiomatic protobuf and how you reference types and nested types in packages. For IDL is the expectation that you use `::` ? I know that the `/` separator is a ROS message-ism so might not be the right approach for IDL.

Does this support concatenated IDL messages? @james-rms can probably point you to some examples on this but my recollection of our rosbag storage mcap support for IDL is that it takes multiple separate IDL definition files and produces a single text with each of these concatenated with some separator. This is similar to how multiple ROS .msg files are concatenated and stored in a schema in mcap. I believe the handling of those concatenated text schemas is also handled within this module during parsing.

File: src/parse.ts [hunk 18]
```diff
@@ -98,6 +108,25 @@ export function fixupTypes(types: MessageDefinition[]): void {
...
+  const parser = new Parser(grammar);
+  parser.feed(messageDefinition);
+  const results = parser.finish();
+
+  if (results.length === 0) {
+    throw new Error(
+      `Could not parse message definition (unexpected end of input): '${messageDefinition}'`,
+    );
+  }
+  const result = results[0] as MessageDefinition[];
```
Comment (defunctzombie) [inline src/parse.ts:121]: Why the cast here? Are we not getting the correct type information from `results`? Is there a way to get correct type information?

File: src/ros2idl.ne [hunk 19]
```diff
@@ -0,0 +1,395 @@
...
+    }
+  });
+  
+  
+  return msgDefs;
+}
+
+function processConstantModule(d) {
+	const moduleName = d[0][4].name;
+  const enclosedConstants = d[0][8];
```
Comment (defunctzombie) [inline src/ros2idl.ne:127]: Some weird spacing happening here?

File: src/ros2idl.ne [hunk 20]
```diff
@@ -0,0 +1,395 @@
...
+const moo = require("moo");
+// Terminal tokens are in all caps
+const lexer = moo.compile({
+  SPACE: {match: /\s+/, lineBreaks: true},
+  DECIMALEXP: /(?:(?:\d+\.\d*)|(?:\d*\.\d+)|(?:[0-9]+))[eE](?:[+|-])?[0-9]+/,
+  DECIMAL: /(?:(?:\d+\.\d*)|(?:\d*\.\d+))/,
+  INTEGER: /[0-9]+/,
+  COMMENT: /(?:\/\/[^\n]*)|(?:\/\*(?:.|\n)+?\*\/)/,
+  HEX_LITERAL: /0x(?:[0-9][a-f][A-F])+?/,
+  STRING: [{match: /"(?:\\["\\rnu]|[^"\\])*?"/, lineBreaks: true, value: x => x.slice(1, -1)}], // remove outside quotes
```
[the remark this answers is not in the document]
Comment (snosenzo) [reply src/ros2idl.ne:51]: ah, good catch. I previously was using multiple matchers, so it's unnecessary. 
`lineBreaks` true is actually unnecessary here, it makes the tokenizer count the number of lineBreaks within it.

File: src/ros2idl.ne [hunk 21]
```diff
@@ -0,0 +1,395 @@
...
+		name: moduleName,
+		definitions: enclosedConstants.flat(1),
+	}];
+}
+
+%}
+
+@lexer lexer
+
+main -> (importDcl __):* module:+ _ {% d => {
```
[the remark this answers is not in the document]
Comment (snosenzo) [reply src/ros2idl.ne:139]: yeah I can update this so that they can be between modules.  though I'm not quite sure we should allow include statements in this context since the schemas in idl should be standalone right? We won't be able to import these so should we disallow them?

File: src/ros2idl.ne [hunk 22]
```diff
@@ -0,0 +1,395 @@
...
+	return d.join("");
+}
+
+// used for combining AST components
+function extend(objs) {
+  return objs.reduce((r, p) => ({ ...r, ...p }), {});
+}
+
+function noop() {
+	return null;
```
[the remark this answers is not in the document]
Comment (snosenzo) [reply src/ros2idl.ne:84]: unfortunately I'm not seeing another way to run prettier on this without making a prettier plugin

[the remark this answers is not in the document]
Comment (defunctzombie) [reply src/ros2idl.ne:84]: Depending on how you want to organize things - nearly does provide a way to use external js files: 
image

Not sure if that ends up being better but it might solve some of your formatting woes

File: src/ros2idl.ne (hunk 7 above)
Comment (snosenzo) [reply src/ros2idl.ne:157]: actually looks like we might just be able to ignore whitespace completely along with comments, if we have the right tokens. from IDL spec:
>Blanks, horizontal and vertical tabs, newlines, form feeds, and comments (collective, "white space") as described
below are ignored except as they serve to separate tokens. Some white space is required to separate otherwise adjacent identifiers, keywords, and constants.

Comment (snosenzo) [reply src/ros2idl.ne:157]: IDL spec states that at least one member is required. 
> The list of all structure members (<member>+) enclosed within braces ({}). Each member (<member>) is defined with a type specification (<type_spec>) followed by a list of at least one declarator (<declarators>). At least one member is required.

File: src/ros2idl.ne (hunk 8 above)
Comment (snosenzo) [reply src/ros2idl.ne:239]: It technically makes it not a _context-free grammar_ but I don't see this causing any issues as is. but we can move it to a postprocessing step if you prefer

File: src/ros2idl.ne (hunk 9 above)
Comment (snosenzo) [reply src/ros2idl.ne:391]: yeah as I commented above the IDL spec does state that we can ignore whitespace. I'm going to try that with the removal of comments in the lexer as well since that's technically also referred to as whitespace in the idl spec.  In that example that removes whitespace I think they might be referring to a tokenizer that doesn't even read whitespace.

File: src/ros2idl.ne (hunk 10 above)
Comment (snosenzo) [reply src/ros2idl.ne:388]: i think that should work, I just originally copied it from the other grammars we have 😅

File: src/ros2idl.ne (hunk 11 above)
Comment (snosenzo) [reply src/ros2idl.ne:382]: converts the token object to string so that it's a little easier to work with. will comment

File: src/parse.ros2idl.test.ts (hunk 15 above)
Comment (snosenzo) [reply src/parse.ros2idl.test.ts:873]: from IDL spec: 
> <module_dcl> ::= "module" <identifier> "{" <definition>+ "}"

File: src/parse.ts (hunk 18 above)
Comment (snosenzo) [reply src/parse.ts:121]: the parser has no knowledge of the output type. This is the same thing we're doing in `buildType` previously

File: src/ros2idl.ne (hunk 21 above)
[the remark this answers is not in the document]
Comment (jtbandes) [reply src/ros2idl.ne:153]: With the format described [here](https://mcap.dev/specification/appendix.html#ros2idl), are we going to see concatenated files that contain includes?

File: src/ros2idl.ne (hunk 5 above)
Comment (jtbandes) [reply src/ros2idl.ne:163]: > using regex to find comments in it will also find comments in strings

Can this be solved by making the string matcher come before the comment matcher in the lexer definition?

Comment (snosenzo) [reply src/ros2idl.ne]: Fixed it by overriding the `.next` on the lexer. comment and whitespace tokens are now entirely ignored

File: src/ros2idl.ne (hunk 6 above)
Comment (snosenzo) [reply src/ros2idl.ne:177]: default values don't apply to structs so we can just ignore all annotations on structs

File: src/parse.ts (hunk 17 above)
Comment (snosenzo) [reply src/parse.ts:26]: I'm more partial to the enums or separate functions, but this is backwards compatible. Don't know what our priorities around this are though

File: src/ros2idl.ne (hunk 21 above)
[the remark this answers is not in the document]
Comment (snosenzo) [reply src/ros2idl.ne]: I think so given the [test case here: ](https://github.com/ros2/rosbag2/blob/8044dba28aa7f99247ccf29861f349969be07117/rosbag2_storage_mcap/test/rosbag2_storage_mcap/test_message_definition_cache.cpp#L63-L126), but they can be ignored by the parser because the definition within the include should be in the same file.

[the remark this answers is not in the document]
Comment (jtbandes) [reply src/ros2idl.ne]: That's what I'd expect yeah. So my question here is: can imports and module definitions ever be interleaved in a way that we'd need to change this to something like `(importDcl | module):+`?

File: src/ros2idl.ne (hunk 8 above)
Comment (jtbandes) [reply src/ros2idl.ne]: Here's a note I found about side effects from the maintainer: https://github.com/kach/nearley/issues/383#issuecomment-400161480

File: src/ros2idl.ne (hunk 21 above)
[the remark this answers is not in the document]
Comment (snosenzo) [reply src/ros2idl.ne]: what I have should allow this now as long as a `module` follows the `#include` ie: modules can't be at the end of files. This makes sense because imported "building blocks" (modules, structes, etc), need to be declared before they are used accored to spec.

File: src/ros2idl.ne (hunk 8 above)
Comment (snosenzo) [reply src/ros2idl.ne]: it looks like it mostly doesn't want you to modify nodes. I think this type of caching should be fine according to this comment by another maintainer further down: https://github.com/kach/nearley/issues/383#issuecomment-400638644

Comment (snosenzo) [reply src/ros2idl.ne]: I think it's fine given that we have a lexer and generally won't be partially reading in this information that will be stored, but if you feel strongly I can hoist it out and have it as a preprocessing step, thought i think that'll be a trickier implementation.

File: src/ros2idl.ne (hunk 16 above)
Comment (snosenzo) [reply src/ros2idl.ne]: updated to read arbitrary annotations

File: src/parse.ts (hunk 17 above)
Comment (snosenzo) [reply src/parse.ts]: updated to have a separate `parseRos2idl` function in `parse.ts`, allowing for only a minor version change

**Comment by snosenzo** [comment]
> What is the correct format for names from idl? For example in protobuf we keep the period separator because that is idiomatic protobuf and how you reference types and nested types in packages. For IDL is the expectation that you use `::` ? I know that the `/` separator is a ROS message-ism so might not be the right approach for IDL.

@defunctzombie So within the file it looks like everything references the `::` name but I'm not sure what the expectation is for displaying/using the ros2idl [package resource name](http://wiki.ros.org/Names#Package_Resource_Names) elsewhere.

File: src/__snapshots__/parse.ros2idl-autoware-large-message.test.ts.snap [hunk 23]
```diff
@@ -7,24 +7,24 @@ Array [
       Object {
         "isComplex": true,
         "name": "header",
-        "type": "std_msgs::msg::Header",
+        "type": "std_msgs/msg::Header",
```
Comment (defunctzombie) [inline src/__snapshots__/parse.ros2idl-autoware-large-message.test.ts.snap:10]: This should be `std_msgs/msg/Header` right?

File: src/__snapshots__/parse.ros2idl-autoware-large-message.test.ts.snap [hunk 24]
```diff
@@ -7,24 +7,24 @@ Array [
...
       },
       Object {
         "arrayUpperBound": undefined,
         "isArray": true,
         "isComplex": true,
         "name": "objects",
-        "type": "autoware_auto_perception_msgs::msg::TrackedObject",
+        "type": "autoware_auto_perception_msgs/msg::TrackedObject",
```
Comment (defunctzombie) [inline src/__snapshots__/parse.ros2idl-autoware-large-message.test.ts.snap:17]: This type still has `::` which should not be present.

File: src/__snapshots__/parse.ros2idl-autoware-large-message.test.ts.snap (hunk 23 above)
Comment (snosenzo) [reply src/__snapshots__/parse.ros2idl-autoware-large-message.test.ts.snap:10]: why wouldn't this test fail?

Comment (snosenzo) [reply src/__snapshots__/parse.ros2idl-autoware-large-message.test.ts.snap:10]: oh wait, my `.replace` only replaces the first -- updating

**Review by defunctzombie — APPROVED** [review]

File: src/parse.ts [hunk 25]
```diff
@@ -106,10 +106,10 @@ export function fixupTypes(types: MessageDefinition[]): void {
  * @returns - parsed message definition
  */
 export function parseRos2idl(messageDefinition: string): MessageDefinition[] {
-  return buildRos2IdlType(messageDefinition, ROS2IDL_GRAMMAR);
+  return buildRos2idlType(messageDefinition, ROS2IDL_GRAMMAR);
```
Comment (defunctzombie) [inline src/parse.ts:109]: Why the change in casing? It does seem like the `i` in `idl` should be capital? Or does lowercase match what we do for `msg` ?

Comment (snosenzo) [reply src/parse.ts:109]: i mostly did it just from `ros2idl`, which is the schema name. I thought it would be more consistent, but I could be convinced otherwise

Comment (jtbandes) [reply src/parse.ts:109]: Let's move this function to a separate file

File: src/parse.ts [hunk 26]
```diff
@@ -12,11 +12,13 @@ import { Grammar, Parser } from "nearley";
...
 import { buildRos2Type } from "./buildRos2Type";
 import ros1Rules from "./ros1.ne";
+import ros2idlRules from "./ros2idl.ne";
 
 const ROS1_GRAMMAR = Grammar.fromCompiled(ros1Rules);
+const ROS2IDL_GRAMMAR = Grammar.fromCompiled(ros2idlRules);
 
 export type ParseOptions = {
-  /** Parse message definitions as ROS2. Otherwise, parse as ROS1 */
+  /** Parse message definitions as ROS 2. Otherwise, parse as ROS1 */
```
Comment (jtbandes) [inline src/parse.ts:21]: OSRF will be very happy

File: src/ros2idl.ne [hunk 27]
```diff
@@ -0,0 +1,395 @@
...
+    return d[0] ? d[0].filter(d => d !== null) : null;
+  }
+%}
+
+annotation -> at %NAME ("(" multiAnnotationParams ")"):? {% d => {
+  const paramsMap = d[2] ? d[2][1] : {};
+  if(d[1].value === "default") {
+    const defaultValue = paramsMap.value;
+    return {defaultValue};
+  }
```
Comment (jtbandes) [inline src/ros2idl.ne:254]: I would suggest moving this out of the grammar and into a postprocessing step, so the grammar remains a fairly simple AST generator

**Review by jtbandes — APPROVED** [review]
My main remaining comment is that I'd suggest removing side effects from the nearley postprocessing functions. I believe they are meant to be side-effect-free and it would feel more clear if we have a simple postprocessing step that gathers the constants & default annotations after running the parser. I'll let you evaluate how difficult those changes would be, it's not a total blocker.

File: package.json [hunk 28]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "@foxglove/rosmsg",
   "version": "4.0.0",
-  "description": "ROS1 and ROS2 message definition parser",
+  "description": "ROS1, ROS2 and ROS2IDL message definition parser",
```
Comment (defunctzombie) [inline package.json:4]: Minor nit: ros2idl is a schema encoding while the other two are not. Technically those are `ros1msg1` and `ros2msg` in the mcap spec: https://mcap.dev/specification/appendix.html#ros1msg

So not sure what this description is trying to say here since ROS 1 and ROS 2 are the frameworks

Comment (jtbandes) [reply package.json:4]: We could say `Parser for ROS .msg and .idl definitions`?

File: src/parse.ts (hunk 26 above)
Comment (amacneil) [reply src/parse.ts:21]: `ROS1` officially outside of OSRF sphere of opinion
